hail/python/hailtop/utils/utils.py

```diff
@@ -258,6 +259,220 @@ async def __aexit__(self,
+        try:
+            await asyncio.wait(tasks)
+        finally:
+            await self._subsema.acquire()
+
+    async def wait_done(self) -> None:
+        while self._pending:
+            if self._exception:
+                raise self._exception
+
+            self._done_event.clear()
+            await self._done_event.wait()
+
+        if self._exception:
+            raise self._exception
+
+    async def __aenter__(self) -> 'OnlineBoundedGather2':
+        await self._subsema.acquire()
+        return self
+
+    async def __aexit__(self,
+                        exc_type: Optional[Type[BaseException]],
+                        exc_val: Optional[BaseException],
+                        exc_tb: Optional[TracebackType]) -> None:
+        self._subsema.release()
+
+        _, exc, _ = sys.exc_info()
+        if exc:
+            await self._shutdown()
+        else:
+            await self.wait_done()
+
+
+async def bounded_gather2_return_exceptions(sema: asyncio.Semaphore, *aws):
+    '''Run the awaitables aws as tasks with parallelism bounded by sema,
+    which should be asyncio.Semaphore whose initial value is the level
+    of parallelism.
+
+    The return value is the list of awaitable results as pairs: the
+    pair (value, None) if the awaitable returned value or (None, exc)
+    if the awaitable raised the exception exc.
+    '''
+    subsema = Subsemaphore(sema)
+
+    async def run_with_sema_return_exceptions(aw):
+        try:
+            async with subsema:
+                return (await aw, None)
+        except:
+            _, exc, _ = sys.exc_info()
+            return (None, exc)
+
+    return await asyncio.gather(*[asyncio.create_task(run_with_sema_return_exceptions(aw)) for aw in aws])
+
+
+async def bounded_gather2_raise_exceptions(sema: asyncio.Semaphore, *aws, cancel_on_error: bool = False):
+    '''Run the awaitables aws as tasks with parallelism bounded by sema,
+    which should be asyncio.Semaphore whose initial value is the level
+    of parallelism.
+
+    The return value is the list of awaitable results.
+
+    The first exception raised by an awaitable is raised by
+    bounded_gather2_raise_exceptions.
+
+    If cancel_on_error is False (the default), the remaining
+    awaitables continue to run with bounded parallelism.  If
+    cancel_on_error is True, the unfinished tasks are all cancelled.
+    '''
+    subsema = Subsemaphore(sema)
+
+    async def run_with_subsema(aw):
+        async with subsema:
+            return await aw
+
+    tasks = [asyncio.create_task(run_with_subsema(aw)) for aw in aws]
+
+    if not cancel_on_error:
+        return await asyncio.gather(*tasks)
+
+    try:
+        return await asyncio.gather(*tasks)
+    finally:
+        _, exc, _ = sys.exc_info()
+        if exc is not None:
+            tasks = []
+            for task in tasks:
+                if not task.done():
+                    task.cancel()
+                    tasks.append(task)
```

Comment:
I don't think this is correct. I don't think you can append to a list you are currently iterating over. I tried it on a small test example and it caused an infinite loop.

Reply:
Yes, this was totally wrong. Fixed.